include/envoy/router/router.h

```diff
@@ -1090,6 +1094,75 @@ class Config {
 };
 
 using ConfigConstSharedPtr = std::shared_ptr<const Config>;
+class GenericConnectionPoolCallbacks;
+class UpstreamRequest;
+class GenericUpstream;
+
+// An API for wrapping either an HTTP or a TCP connection pool.
+class GenericConnPool : public Logger::Loggable<Logger::Id::router> {
+public:
+  virtual ~GenericConnPool() = default;
+
+  // Called to create a new HTTP stream or TCP connection. The implementation
+  // is then responsible for calling either onPoolReady or onPoolFailure on the
+  // supplied GenericConnectionPoolCallbacks.
+  virtual void newStream(GenericConnectionPoolCallbacks* callbacks) PURE;
+  // Called to cancel a call to newStream. Returns true if a newStream request
+  // was canceled, false otherwise.
+  virtual bool cancelAnyPendingRequest() PURE;
```

Comment:
From an API perspective I would recommend having `newStream()` return a handle which can be cancelled? I think that would be cleaner and more future proof.

Reply:
I think if you're OK with it I'll do one PR just moving the APIs out and cleaning them up, then a follow-up doing factory and moving code around.  I think if we do both in the same PR it's hard to see what's changing.